kore/src/Kore/Equation/Application.hs

```diff
@@ -174,6 +179,28 @@ attemptEquation sideCondition termLike equation =
         matchIncremental term1 term2
         & MaybeT & noteT matchError
 
+    applyAndSelectMatchResult
+        :: [MatchResult (Target variable)]
+        -> ExceptT
+            (AttemptEquationError variable)
+            simplifier
+            (Equation variable, Predicate variable)
+    applyAndSelectMatchResult [] =
+        whileApplyMatchResult
+        $ applyMatchResult
+            equationRenamed
+            (Predicate.makeFalsePredicate_, mempty)
+    applyAndSelectMatchResult results =
+        Foldable.foldr1
+            takeFirstSuccess
+            ( whileApplyMatchResult
+            . applyMatchResult equationRenamed
+            <$> results
+            )
+    takeFirstSuccess first second = do
+        lift (runExceptT first)
+        >>= either (const second) (const first)
```

Comment:
This is not quite right: it will call `first` twice if it succeeds. What we really want to do is:
```suggestion
    takeFirstSuccess first second = catchError first (const second)
```